### PR TITLE
ALTAPPS-1115: Shared paywall feature

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[ALTAPPS-1115](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-1115)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] View analytics events are added for new screens;
- [x] New analytics events are documented;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Implement `PaywallFeature`
- Open Paywall from `WelcomeOnboardingFeature` for users with freemium subscription

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt [hunk 1]
```diff
@@ -255,15 +254,10 @@ internal class AppReducer(
         finishAction: WelcomeOnboardingFeature.Action.OnboardingFlowFinished
     ): Set<Action> =
         setOf(
-            when (val reason = finishAction.reason) {
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt:258]: JFYI I understand that the code works as expected (like before), but in my opinion _finish reason_ previously simplified readability and understanding of the code

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt:258]: @ivan-magda 
Previously, `TrackSelectionScreen` was shown only for users who passed NotificationOnboarding. Now also we should show TrackSelection for those, who have seen Paywall, but only if they are new users.
I guess, at the moment, it is easier to lean only on the profile.

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt [hunk 2]
```diff
@@ -158,7 +157,7 @@ internal class AppReducer(
             )
             val (onboardingState, onboardingActions) = reduceWelcomeOnboardingMessage(
                 WelcomeOnboardingFeature.State(),
-                WelcomeOnboardingFeature.InternalMessage.OnboardingFlowRequested(
+                WelcomeOnboardingFeature.Message.OnboardingFlowRequested(
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt:160]: You are right, we should leave `OnboardingFlowRequested ` as an internal message, to not allow its usage outside of the shared module.

File: shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt:1]: @ivan-magda 
I've added `PaywallTransitionSource` to use in the analytical events

**Comment by XanderZhu** [comment]
@ivan-magda 
I've added subscription price fetching

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt (hunk 1 above)
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppReducer.kt:258]: OK

File: shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt [hunk 3]
```diff
@@ -15,15 +16,26 @@ import org.hyperskill.app.analytic.domain.model.hyperskill.HyperskillAnalyticTar
...
     HyperskillAnalyticAction.CLICK,
     HyperskillAnalyticPart.MAIN,
     HyperskillAnalyticTarget.BUY_SUBSCRIPTION
-)
\ No newline at end of file
+) {
+    override val params: Map<String, Any>
+        get() = super.params + mapOf(
+            PaywallAnalyticKeys.PAYWALL_TRANSITION_SOURCE to paywallTransitionSource.analyticName
+        )
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt:40]: To put value to the `context` you need to use a different syntax:
```kotlin
override val params: Map<String, Any>
    get() = super.params +
        mapOf(
            PARAM_CONTEXT to mapOf(
                PaywallAnalyticKeys.PAYWALL_TRANSITION_SOURCE to paywallTransitionSource.analyticName
            )
        )
```

Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt:40]: please also update all paywall analytic events `params` computed properties

**Review by ivan-magda — APPROVED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt (hunk 3 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/paywall/domain/analytic/PaywallClickedBuySubscriptionHyperskillAnalyticEvent.kt:40]: @ivan-magda 
I suggest consider a special property, like `contextParams: Map<String, Any>`, to easier add context params

**Pull request merged**